File: components/master/tools/policy/tokenize.ts
// components/master/tools/policy/tokenize.ts
//
// Deterministic shell-aware tokenizer for the subctl policy engine (v2.7.0).
//
// Per `.orchestration/handoff-pack/06-tool-family-policy.md` §4 + pack 11 §2.1:
//
//   "The tokenizer is the single most important piece of correctness in this
//    file. The exact same tokenization must happen in the master daemon AND
//    in the CLI hook, or denials and allows will diverge. Use one shared
//    module imported by both call sites. Recommended: `shell-quote` from npm,
//    which is small and battle-tested. Pin the version."
//
// This file is the SPEC for PR 8's Go port (`bin/subctl-policy-check`). Every
// vector in `config/policy/test-vectors.toml` must produce identical token
// streams in both implementations. If you change anything here, also update
// the Go side or `vectors.test.ts` will diverge.
//
// Determinism contract:
//   - tokenize(s) is a pure function of `s`.
//   - 1000 trials on the same input produce byte-identical arrays
//     (`tokenize.test.ts` asserts this explicitly).
//   - No process-time, no random, no `Date.now()`, no env reads.
//
// Expansion contract (pack 06 §4 "no shell expansion"):
//   - `$VAR` / `${VAR}` → kept LITERALLY as `$VAR`.
//   - `~/foo` → kept LITERALLY as `~/foo`.
//   - `*.txt` / `?` globs → kept LITERALLY as the source pattern.
//   - The whole point is that `rm -rf $HOME` still has the substring
//     `rm -rf`, so deny_always.substrings can catch it.
//
// Operator preservation (pack 11 §2.1):
//   - `|`, `||`, `&&`, `&`, `;`, `>`, `>>`, `<`, `<<<` etc. are emitted as
//     their literal string. They appear as separate tokens so check.ts can
//     reason about pipelines (though deny_always.regex on the RAW command
//     string is the dominant pipeline-deny path; see check.ts).
//   - Heredoc: shell-quote splits `<<EOF` into two `<` ops followed by the
//     tag string `EOF`. We merge them into a single `<<EOF` token so that
//     `python <<EOF` tokenizes as `["python", "<<EOF", …]` per pack 11 §2.1.

import { parse } from "shell-quote";

/**
 * Tokenize a shell command line into an array of literal-string tokens.
 *
 * Empty input and whitespace-only input both return `[]`. shell-quote errors
 * (which are rare; the parser is permissive) collapse to `[]` so the hot path
 * falls into default_deny rather than crashing the worker.
 *
 * NOTE: This function does NOT decide allow/deny; that's check.ts. It only
 * converts a raw command string into an array of string tokens. The literal-
 * string contract means `tokens.some(t => t.includes(substring))` and
 * `firstNonFlag = tokens.find(t => !t.startsWith("-"))` work the way pack 06
 * §4 expects.
 */
export function tokenize(cmd: string): string[] {
  if (!cmd) return [];
  // shell-quote returns [] for an all-whitespace input on its own, but we
  // short-circuit so the function has a deterministic fast path.
  if (cmd.trim() === "") return [];

  // Tell shell-quote to keep variables literal. Without this, `$HOME` would
  // become "" (empty token) — fatal, because then `rm -rf $HOME` would
  // tokenize as `["rm", "-rf", ""]` and the empty token would silently pass
  // the deny check. Passing this lambda makes `$HOME` stay as the literal
  // string "$HOME" — exactly what pack 06 §4 mandates.
  const keepLiteral = (name: string) => `$${name}`;

  let parsed: unknown[];
  try {
    parsed = parse(cmd, keepLiteral as unknown as Record<string, string>);
  } catch {
    return [];
  }

  const out: string[] = [];
  for (let i = 0; i < parsed.length; i++) {
    const item = parsed[i];
    if (typeof item === "string") {
      out.push(item);
      continue;
    }
    if (item && typeof item === "object" && "op" in item) {
      const op = String((item as { op: string }).op);

      // Glob entries arrive as `{ op: 'glob', pattern: '*.txt' }`. The
      // expansion contract says we keep the literal pattern.
      if (op === "glob") {
        const pat = (item as { pattern?: string }).pattern;
        if (typeof pat === "string") out.push(pat);
        continue;
      }

      // Heredoc handling: shell-quote splits `<<` into two consecutive
      // `{op:'<'}` entries. Combine them with the following tag string into
      // a single token like `<<EOF`, per pack 11 §2.1 ("Heredocs: python
      // <<EOF\n…EOF → tokenized as python, <<EOF, …").
      const nextItem = parsed[i + 1];
      const isNextLT =
        nextItem != null &&
        typeof nextItem === "object" &&
        "op" in nextItem &&
        (nextItem as { op?: string }).op === "<";
      if (op === "<" && isNextLT) {
        const tagItem = parsed[i + 2];
        if (typeof tagItem === "string") {
          out.push(`<<${tagItem}`);
          i += 2;
          continue;
        }
        // No tag string after the `<<`: emit as bare `<<` and continue.
        out.push("<<");
        i += 1;
        continue;
      }

      // Every other operator: emit as a single literal-string token
      // (`|` → "|", `&&` → "&&", `<<<` → "<<<", etc.).
      out.push(op);
    }
    // Anything else (e.g. shell-quote's `{ comment: "..." }`): drop silently.
  }
  return out;
}

File: bin/subctl-policy-check/tokenize.go
// bin/subctl-policy-check/tokenize.go
//
// Deterministic shell-aware tokenizer for the policy engine's Go port.
//
// Spec: `components/master/tools/policy/tokenize.ts`. The TS implementation
// wraps the npm `shell-quote` library; this file ports the relevant subset of
// shell-quote's behavior. Determinism contract (pack 11 §2.1): tokenize(s) is
// a pure function of `s`; the same input produces byte-identical token arrays
// across both implementations.
//
// Expansion contract (pack 06 §4 "no shell expansion"):
//   - `$VAR` / `${VAR}` → kept LITERALLY as `$VAR` / `${VAR}`.
//   - `~/foo` → kept LITERALLY as `~/foo`.
//   - `*.txt` / `?` globs → kept LITERALLY as the source pattern.
//   - The whole point is that `rm -rf $HOME` still has the substring
//     `rm -rf`, so deny_always.substrings can catch it. Decisions are made
//     on the raw command line as well, so even if tokenization differs at
//     the margins, substring-level denies fire first.
//
// Operator preservation (pack 11 §2.1):
//   - `|`, `||`, `&&`, `&`, `;`, `>`, `>>`, `<`, `<<<` → emitted as their
//     literal string in their own token slot.
//   - Heredoc: shell-quote splits `<<EOF` into two `<` ops plus the tag
//     string. tokenize.ts merges them to a single `<<EOF` token. We do the
//     same so the regex denies fire on the merged form.

package main

import "strings"

// tokenize converts a raw command line into a flat slice of literal-string
// tokens. Empty input and whitespace-only input both return nil.
//
// Behavior summary:
//   - whitespace separates tokens
//   - single quotes preserve everything literally
//   - double quotes preserve everything except `\\`, `\"`, `\$`, `` \` ``
//     (which collapse to the escaped char). This matches shell-quote's
//     escape handling in JS.
//   - backslash outside quotes escapes the next character (including space)
//   - `|`, `||`, `&`, `&&`, `;`, `>`, `>>`, `<` are emitted as their own tokens
//   - `<<<` is emitted as `<<<`
//   - `<<TAG` is emitted as a single `<<TAG` token (heredoc merge)
//
// The function never panics; malformed input (e.g. unterminated quote) is
// handled best-effort by consuming what's there and emitting it.
func tokenize(cmd string) []string {
	if strings.TrimSpace(cmd) == "" {
		return nil
	}

	var tokens []string
	var cur strings.Builder
	// hasToken tracks whether the current token-in-progress is "real" — i.e.
	// the cursor entered a quote or wrote a char — so an empty `""` token
	// flushes correctly. We don't currently rely on it for our 76 vectors but
	// it keeps the contract honest.
	hasToken := false

	flush := func() {
		if hasToken {
			tokens = append(tokens, cur.String())
			cur.Reset()
			hasToken = false
		}
	}

	runes := []rune(cmd)
	n := len(runes)

	for i := 0; i < n; {
		c := runes[i]
		switch {
		case c == ' ' || c == '\t' || c == '\n' || c == '\r':
			flush()
			i++

		case c == '\'':
			// Single quote: literal everything until next single quote. No
			// escapes are recognized inside (matches shell + shell-quote).
			hasToken = true
			i++
			for i < n && runes[i] != '\'' {
				cur.WriteRune(runes[i])
				i++
			}
			if i < n {
				i++ // skip closing '
			}

		case c == '"':
			// Double quote: literal except for `\\`, `\"`, `\$`, `` \` ``
			// (shell-quote's recognized escapes). Unrecognized backslash
			// sequences pass through with the backslash.
			hasToken = true
			i++
			for i < n && runes[i] != '"' {
				if runes[i] == '\\' && i+1 < n {
					next := runes[i+1]
					switch next {
					case '\\', '"', '$', '`':
						cur.WriteRune(next)
						i += 2
						continue
					}
				}
				cur.WriteRune(runes[i])
				i++
			}
			if i < n {
				i++ // skip closing "
			}

		case c == '\\':
			// Backslash outside quotes: literal next char (incl. whitespace).
			if i+1 < n {
				cur.WriteRune(runes[i+1])
				hasToken = true
				i += 2
			} else {
				// Trailing backslash with no follow — drop it.
				i++
			}

		case c == '|':
			flush()
			if i+1 < n && runes[i+1] == '|' {
				tokens = append(tokens, "||")
				i += 2
			} else {
				tokens = append(tokens, "|")
				i++
			}

		case c == '&':
			flush()
			if i+1 < n && runes[i+1] == '&' {
				tokens = append(tokens, "&&")
				i += 2
			} else {
				tokens = append(tokens, "&")
				i++
			}

		case c == ';':
			flush()
			tokens = append(tokens, ";")
			i++

		case c == '>':
			flush()
			if i+1 < n && runes[i+1] == '>' {
				tokens = append(tokens, ">>")
				i += 2
			} else {
				tokens = append(tokens, ">")
				i++
			}

		case c == '<':
			flush()
			// `<<<` is here-string.
			if i+2 < n && runes[i+1] == '<' && runes[i+2] == '<' {
				tokens = append(tokens, "<<<")
				i += 3
				break
			}
			// `<<TAG` is a heredoc start. shell-quote splits this into two
			// `<` ops + a string tag; tokenize.ts merges them. We emit the
			// already-merged form so the regex denies (`python3?\s*<<\s*EOF`)
			// still fire against the raw command line, and so the token list
			// carries a meaningful `<<EOF` slot.
			if i+1 < n && runes[i+1] == '<' {
				i += 2
				// shell-quote merges only when a string tag immediately
				// follows (no whitespace). We mirror that.
				var tag strings.Builder
				for i < n {
					ch := runes[i]
					if ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r' {
						break
					}
					if ch == '|' || ch == '&' || ch == ';' || ch == '<' || ch == '>' {
						break
					}
					tag.WriteRune(ch)
					i++
				}
				if tag.Len() > 0 {
					tokens = append(tokens, "<<"+tag.String())
				} else {
					tokens = append(tokens, "<<")
				}
				break
			}
			tokens = append(tokens, "<")
			i++

		default:
			cur.WriteRune(c)
			hasToken = true
			i++
		}
	}
	flush()
	return tokens
}
